File: java/core/src/test/java/dev/varar/core/CellDiffTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/cell-diff.test.ts}. */
class CellDiffTest {

    private static final Span SPAN = new Span(0, 1, 1, 1, 1, 2);
    private static final List<CellDiff.RowCheck> CHECKS =
            List.of(new CellDiff.RowCheck("dice", "3, 3, 3, 4, 4", SPAN), new CellDiff.RowCheck("score", "9", SPAN));

    @Test
    void aReturnedColumnThatMatchesItsCellIsOk() {
        List<CellDiff> diffs = CellDiff.compareRow(Map.of("score", 9), CHECKS);
        assertEquals(List.of(new CellDiff("score", SPAN, "9", "9", true)), diffs);
    }

    @Test
    void aReturnedColumnThatDiffersIsNotOkWithExpectedAndActual() {
        List<CellDiff> diffs = CellDiff.compareRow(Map.of("score", 6), CHECKS);
        assertEquals(List.of(new CellDiff("score", SPAN, "9", "6", false)), diffs);
    }

    @Test
    void columnsThatAreNotReturnedAreInputsNotChecked() {
        // `dice` is never returned, so it never appears in the diffs.
        assertEquals(
                List.of("score"),
                CellDiff.compareRow(Map.of("score", 9), CHECKS).stream()
                        .map(CellDiff::column)
                        .toList());
    }

    @Test
    void aReturnedKeyThatIsNotAColumnIsIgnored() {
        assertEquals(List.of(), CellDiff.compareRow(Map.of("nope", 1), CHECKS));
    }

    @Test
    void nullNonMapReturnChecksNothing() {
        assertEquals(List.of(), CellDiff.compareRow(null, CHECKS));
        assertEquals(List.of(), CellDiff.compareRow(42, CHECKS));
    }

    @Test
    void cellMismatchExceptionCarriesTheCellsAndIsDetectable() {
        CellDiff.CellMismatchException err =
                new CellDiff.CellMismatchException(List.of(new CellDiff("score", SPAN, "9", "6", false)));
        assertTrue(CellDiff.isCellMismatchException(err));
        assertFalse(CellDiff.isCellMismatchException(new RuntimeException("x")));
        assertEquals("6", err.cells().get(0).actual());
        assertTrue(err.getMessage().contains("score"));
    }

    // Build a real Table (with cellSpans) by parsing a markdown table.
    private record ParsedTable(Ast.Table table, String source) {}

    private static ParsedTable tableOf(String source) {
        Ast.VarDoc doc = Parse.parse("t.md", source);
        Ast.Table table = doc.examples().get(0).body().stream()
                .filter(b -> b instanceof Ast.Table)
                .map(b -> (Ast.Table) b)
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("no table parsed"));
        return new ParsedTable(table, source);
    }

    private static final String TABLE_SRC = """
            # T

            these:

            | before | after |
            | ------ | ----- |
            | var    | VAR   |
            | bdd    | BDD   |""";

    @Test
    void compareTableArrayOfArraysFullMatchAllOk() {
        Ast.Table table = tableOf(TABLE_SRC).table();
        List<CellDiff> diffs = CellDiff.compareTable(List.of(List.of("var", "VAR"), List.of("bdd", "BDD")), table);
        assertEquals(4, diffs.size());
        assertTrue(diffs.stream().allMatch(CellDiff::ok));
    }

    @Test
    void compareTableArrayOfRecordsFullMatchAllOk() {
        Ast.Table table = tableOf(TABLE_SRC).table();
        List<CellDiff> diffs = CellDiff.compareTable(
                List.of(Map.of("before", "var", "after", "VAR"), Map.of("before", "bdd", "after", "BDD")), table);
        assertTrue(diffs.stream().allMatch(CellDiff::ok));
    }

    @Test
    void compareTableOneWrongCellNotOkWithExpectedActualSpan() {
        ParsedTable parsed = tableOf(TABLE_SRC);
        List<CellDiff> diffs =
                CellDiff.compareTable(List.of(List.of("var", "WRONG"), List.of("bdd", "BDD")), parsed.table());
        List<CellDiff> bad = diffs.stream().filter(d -> !d.ok()).toList();
        assertEquals(1, bad.size());
        assertEquals("after", bad.get(0).column());
        assertEquals("VAR", bad.get(0).expected());
        assertEquals("WRONG", bad.get(0).actual());
        assertEquals(
                "VAR",
                parsed.source()
                        .substring(
                                bad.get(0).span().startOffset(),
                                bad.get(0).span().endOffset()));
    }

    @Test
    void compareTableNumbersAreStringifiedBeforeCompare() {
        Ast.Table table = tableOf("""
                                # T

                                these:

                                | n |
                                | - |
                                | 7 |""").table();
        assertTrue(CellDiff.compareTable(List.of(List.of(7)), table).stream().allMatch(CellDiff::ok));
    }

    @Test
    void compareTableNullReturnChecksNothing() {
        Ast.Table table = tableOf(TABLE_SRC).table();
        assertEquals(List.of(), CellDiff.compareTable(null, table));
    }

    @Test
    void compareTableExtraKeysOnAReturnedRecordAreIgnored() {
        Ast.Table table = tableOf(TABLE_SRC).table();
        List<CellDiff> diffs = CellDiff.compareTable(
                List.of(
                        Map.of("before", "var", "after", "VAR", "extra", "ignored"),
                        Map.of("before", "bdd", "after", "BDD", "note", 123)),
                table);
        assertTrue(diffs.stream().allMatch(CellDiff::ok));
        assertEquals(
                List.of("before", "after", "before", "after"),
                diffs.stream().map(CellDiff::column).toList());
    }

    @Test
    void compareTableShapeTypeErrorsThrowReturnShapeException() {
        Ast.Table table = tableOf(TABLE_SRC).table();
        assertThrows(CellDiff.ReturnShapeException.class, () -> CellDiff.compareTable("nope", table)); // not a list
        assertThrows(
                CellDiff.ReturnShapeException.class,
                () -> CellDiff.compareTable(List.of(List.of("var", "VAR")), table)); // wrong row count
        assertThrows(
                CellDiff.ReturnShapeException.class,
                () -> CellDiff.compareTable(List.of(List.of("var"), List.of("bdd")), table)); // wrong width
        assertThrows(
                CellDiff.ReturnShapeException.class,
                () -> CellDiff.compareTable(
                        List.of(Map.of("before", "var"), Map.of("before", "bdd")), table)); // missing key
        assertThrows(
                CellDiff.ReturnShapeException.class,
                () -> CellDiff.compareTable(
                        List.of(List.of("var", "VAR"), Map.of("before", "bdd", "after", "BDD")), table)); // mixed forms
    }
}

File: rust/core/tests/cell_diff_test.rs
//! Port of `CellDiffTest.java` / `cell-diff.test.ts`.

mod common;

use common::{vlist, vmap};
use varar_core::ast::{Block, Table};
use varar_core::cell_diff::{CellDiff, RowCheck, compare_row, compare_table};
use varar_core::error::StepError;
use varar_core::offsets::utf16_slice;
use varar_core::parse::parse;
use varar_core::span::Span;
use varar_core::value::Value;

const SPAN: Span = Span {
    start_offset: 0,
    end_offset: 1,
    start_line: 1,
    start_col: 1,
    end_line: 1,
    end_col: 2,
};

fn checks() -> Vec<RowCheck> {
    vec![
        RowCheck::new("dice", "3, 3, 3, 4, 4", SPAN),
        RowCheck::new("score", "9", SPAN),
    ]
}

const TABLE_SRC: &str = "# T\n\nthese:\n\n| before | after |\n| ------ | ----- |\n| var    | VAR   |\n| bdd    | BDD   |";

fn table_of(source: &str) -> Table {
    let doc = parse("t.md", source);
    doc.examples[0]
        .body
        .iter()
        .find_map(|b| {
            if let Block::Table(t) = b {
                Some(t.clone())
            } else {
                None
            }
        })
        .expect("no table parsed")
}

#[test]
fn a_returned_column_that_matches_its_cell_is_ok() {
    let diffs = compare_row(Some(&vmap(vec![("score", Value::Int(9))])), &checks());
    assert_eq!(vec![CellDiff::new("score", SPAN, "9", "9", true)], diffs);
}

#[test]
fn a_returned_column_that_differs_is_not_ok_with_expected_and_actual() {
    let diffs = compare_row(Some(&vmap(vec![("score", Value::Int(6))])), &checks());
    assert_eq!(vec![CellDiff::new("score", SPAN, "9", "6", false)], diffs);
}

#[test]
fn columns_that_are_not_returned_are_inputs_not_checked() {
    let diffs = compare_row(Some(&vmap(vec![("score", Value::Int(9))])), &checks());
    let cols: Vec<String> = diffs.iter().map(|d| d.column.clone()).collect();
    assert_eq!(vec!["score".to_string()], cols);
}

#[test]
fn a_returned_key_that_is_not_a_column_is_ignored() {
    assert_eq!(
        Vec::<CellDiff>::new(),
        compare_row(Some(&vmap(vec![("nope", Value::Int(1))])), &checks())
    );
}

#[test]
fn null_non_map_return_checks_nothing() {
    assert_eq!(Vec::<CellDiff>::new(), compare_row(None, &checks()));
    assert_eq!(Vec::<CellDiff>::new(), compare_row(Some(&Value::Int(42)), &checks()));
}

#[test]
fn cell_mismatch_carries_the_cells_and_is_detectable() {
    let err = StepError::CellMismatch(vec![CellDiff::new("score", SPAN, "9", "6", false)]);
    assert!(err.as_cell_mismatch().is_some());
    assert!(
        StepError::Handler(varar_core::error::HandlerError::new("x"))
            .as_cell_mismatch()
            .is_none()
    );
    assert_eq!("6", err.as_cell_mismatch().unwrap()[0].actual);
    assert!(err.message().contains("score"));
}

#[test]
fn compare_table_array_of_arrays_full_match_all_ok() {
    let table = table_of(TABLE_SRC);
    let diffs = compare_table(
        Some(&vlist(vec![
            vlist(vec![Value::from("var"), Value::from("VAR")]),
            vlist(vec![Value::from("bdd"), Value::from("BDD")]),
        ])),
        &table,
    )
    .unwrap();
    assert_eq!(4, diffs.len());
    assert!(diffs.iter().all(|d| d.ok));
}

#[test]
fn compare_table_array_of_records_full_match_all_ok() {
    let table = table_of(TABLE_SRC);
    let diffs = compare_table(
        Some(&vlist(vec![
            vmap(vec![
                ("before", Value::from("var")),
                ("after", Value::from("VAR")),
            ]),
            vmap(vec![
                ("before", Value::from("bdd")),
                ("after", Value::from("BDD")),
            ]),
        ])),
        &table,
    )
    .unwrap();
    assert!(diffs.iter().all(|d| d.ok));
}

#[test]
fn compare_table_one_wrong_cell_not_ok_with_expected_actual_span() {
    let table = table_of(TABLE_SRC);
    let diffs = compare_table(
        Some(&vlist(vec![
            vlist(vec![Value::from("var"), Value::from("WRONG")]),
            vlist(vec![Value::from("bdd"), Value::from("BDD")]),
        ])),
        &table,
    )
    .unwrap();
    let bad: Vec<&CellDiff> = diffs.iter().filter(|d| !d.ok).collect();
    assert_eq!(1, bad.len());
    assert_eq!("after", bad[0].column);
    assert_eq!("VAR", bad[0].expected);
    assert_eq!("WRONG", bad[0].actual);
    assert_eq!("VAR", utf16_slice(TABLE_SRC, bad[0].span.start_offset, bad[0].span.end_offset));
}

#[test]
fn compare_table_numbers_are_stringified_before_compare() {
    let table = table_of("# T\n\nthese:\n\n| n |\n| - |\n| 7 |");
    let diffs = compare_table(Some(&vlist(vec![vlist(vec![Value::Int(7)])])), &table).unwrap();
    assert!(diffs.iter().all(|d| d.ok));
}

#[test]
fn compare_table_null_return_checks_nothing() {
    let table = table_of(TABLE_SRC);
    assert_eq!(Vec::<CellDiff>::new(), compare_table(None, &table).unwrap());
}

#[test]
fn compare_table_extra_keys_on_a_returned_record_are_ignored() {
    let table = table_of(TABLE_SRC);
    let diffs = compare_table(
        Some(&vlist(vec![
            vmap(vec![
                ("before", Value::from("var")),
                ("after", Value::from("VAR")),
                ("extra", Value::from("ignored")),
            ]),
            vmap(vec![
                ("before", Value::from("bdd")),
                ("after", Value::from("BDD")),
                ("note", Value::Int(123)),
            ]),
        ])),
        &table,
    )
    .unwrap();
    assert!(diffs.iter().all(|d| d.ok));
    let cols: Vec<String> = diffs.iter().map(|d| d.column.clone()).collect();
    assert_eq!(vec!["before", "after", "before", "after"], cols);
}

#[test]
fn compare_table_shape_type_errors_throw_return_shape() {
    let table = table_of(TABLE_SRC);
    let is_shape =
        |r: Result<Vec<CellDiff>, StepError>| matches!(r.unwrap_err(), StepError::ReturnShape(_));
    assert!(is_shape(compare_table(Some(&Value::from("nope")), &table))); // not a list
    assert!(is_shape(compare_table(
        Some(&vlist(vec![vlist(vec![Value::from("var"), Value::from("VAR")])])),
        &table
    ))); // wrong row count
    assert!(is_shape(compare_table(
        Some(&vlist(vec![
            vlist(vec![Value::from("var")]),
            vlist(vec![Value::from("bdd")])
        ])),
        &table
    ))); // wrong width
    assert!(is_shape(compare_table(
        Some(&vlist(vec![
            vmap(vec![("before", Value::from("var"))]),
            vmap(vec![("before", Value::from("bdd"))]),
        ])),
        &table
    ))); // missing key
    assert!(is_shape(compare_table(
        Some(&vlist(vec![
            vlist(vec![Value::from("var"), Value::from("VAR")]),
            vmap(vec![
                ("before", Value::from("bdd")),
                ("after", Value::from("BDD"))
            ]),
        ])),
        &table
    ))); // mixed forms
}
